hack/promqlrule.go

```diff
@@ -0,0 +1,75 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+
+	promql "github.com/prometheus/prometheus/promql/parser"
+)
+
+func extractMetricNames(file *os.File) (map[string]bool, error) {
+	metricNamesMap := map[string]bool{}
+	r := bufio.NewReader(file)
+	for {
+		line, _, err := r.ReadLine()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+
+		var lineStr string
+
+		json.Unmarshal(line, &lineStr)
+
+		expr, err := promql.ParseExpr(lineStr)
+
+		if err != nil {
+			return nil, err
+		}
+
+		selectors := promql.ExtractSelectors(expr)
+		for _, selector := range selectors {
+			for _, label := range selector {
+				if label.Name == "__name__" {
+					metricNamesMap[label.Value] = true
+				} else {
+					metricNamesMap[label.Name] = true
+				}
+			}
+		}
```

Comment:
I don't that this will yield exactly what you are expecting. Instead of just recording metrics, you'll also get some extract labels.
It would be more appropriate to have the following piece if you only want to record metric names:
```go
promql.Inspect(expr, func(node Node, _ []Node) error {
  vs, ok := node.(*promql.VectorSelector)
  if ok {
    metricNameMap[vs.Name] = struct{}{}
  }
  return nil
})
```

Reply:
Updated now, ready to review again :)